beacon-chain/cache/attestation.go

```diff
@@ -0,0 +1,262 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/operations/attestations/forkchoice"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1/attestation"
+)
+
+type attGroup struct {
+	slot     primitives.Slot
+	local    ethpb.Att
+	external []ethpb.Att
+}
+
+type AttestationCache struct {
+	attestations map[attestation.Id]*attGroup
+	sync.RWMutex
+	forkchoiceAtts *forkchoice.Attestations
+}
+
+func NewAttestationCache() *AttestationCache {
+	return &AttestationCache{
+		attestations:   make(map[attestation.Id]*attGroup),
+		forkchoiceAtts: forkchoice.New(),
+	}
+}
+
+func (c *AttestationCache) Add(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.attestations[id]
+	if group == nil {
+		group = &attGroup{
+			slot: att.GetData().Slot,
+		}
+		c.attestations[id] = group
+	}
+
+	if att.IsAggregated() {
+		group.external = append(group.external, att.Clone())
+		return nil
+	}
+
+	local := c.attestations[id].local
+	if local == nil {
+		local = att.Clone()
+	}
+	bit := att.GetAggregationBits().BitIndices()[0]
+	if local.GetAggregationBits().BitAt(uint64(bit)) {
+		return nil
+	}
+	sig, err := aggregateSig(local, att)
+	if err != nil {
+		return errors.Wrapf(err, "could not aggregate signatures")
+	}
+
+	local.GetAggregationBits().SetBitAt(uint64(bit), true)
+	local.SetSignature(sig)
+
+	return nil
+}
+
+func (c *AttestationCache) GetAll() []ethpb.Att {
+	c.RLock()
+	defer c.RUnlock()
+
+	var result []ethpb.Att
+	for _, group := range c.attestations {
+		if group.local != nil {
+			result = append(result, group.local)
+		}
+		for _, a := range group.external {
+			result = append(result, a)
+		}
+	}
+	return result
+}
+
+func (c *AttestationCache) Count() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	count := 0
+	for _, group := range c.attestations {
+		if group.local != nil {
+			count++
+		}
+		count += len(group.external)
+	}
+	return count
+}
+
+func (c *AttestationCache) DeleteCovered(att ethpb.Att) error {
```

Comment:
`DeleteCovered` is much more accurate than the current `DeleteAggregateAttestation`. The current name is misleading because we do not delete the passed attestation, but all attestations in the pool whose bits are covered by the incoming attestation.